hivemind/client/optim/simple.py

```diff
@@ -0,0 +1,88 @@
+from threading import Thread, Lock, Event
+from typing import Optional, Callable, Any
+
+import torch
+
+from hivemind.dht import DHT
+from hivemind.client.averaging import DecentralizedAverager
+from hivemind.client.optim.base import DecentralizedOptimizerBase
+from hivemind.utils import MPFuture, get_logger
+
+logger = get_logger(__name__)
+
+
+class DecentralizedOptimizer(DecentralizedOptimizerBase):
+    """
+    A simple optimizer that trains a shared model by averaging model parameters with peers in the background.
+
+    :param opt: a pytorch optimizer configured to update model parameters.
+    :param dht: a running hivemind DHT daemon connected to other peers
+    :param prefix: all DHT keys that point to optimization metadata will have this prefix
+    :param target_group_size: maximum group size for averaging (see DecentralizedAverager)
+    :param kwargs: additional parameters passed to DecentralizedAverager
+
+    :note: using DecentralizedOptimizer with adaptive learning rates may result in poor convergence due to
+      out-of-sync adaptive learning rates (such as adam second momentum or schedule step). Please ensure that these
+      statistics are synchronized or use a more advanced DecentralizedOptimizer version, if applicable.
+    :note: the base optimizer cannot add param groups after the DecentralizedOptimizer is created
+    """
+    averager_step: Optional[MPFuture] = None
+
+    def __init__(self, opt: torch.optim.Optimizer, dht: DHT, prefix: str, target_group_size: int,
+                 timeout: Optional[float] = None, verbose: bool = False, **kwargs):
+        super().__init__(opt, dht)
+        with torch.no_grad():
+            averaged_tensors = tuple(p.cpu().float().clone().requires_grad_(False)
+                                     for group in self.param_groups for p in group['params'])
+        self.averager = DecentralizedAverager(averaged_tensors, dht, start=True, prefix=prefix,
+                                              target_group_size=target_group_size, **kwargs)
+        self.lock_parameters, self.update_event, self.stop_event = Lock(), Event(), Event()
+        self.background_averaging_thread = Thread(daemon=True, target=_average_parameters_in_background,
+                                                  args=[self.update_event, self.stop_event, self.averager, self.opt],
+                                                  kwargs=dict(verbose=verbose, timeout=timeout))
+        self.background_averaging_thread.start()
+
+    def step(self, closure: Optional[Callable[[], Any]] = None):
+        self.update_event.set()
+        with self.lock_parameters:
+            return self.opt.step(closure)
+
+    def zero_grad(self, set_to_none: Optional[bool] = False):
+        return self.opt.zero_grad()
```

Comment:
`set_to_none` isn't passed to the original optimizer

Reply:
nice catch!